### DISCUSSION TITLE
Docker-based Dokku deployment with multiple processes in one app

### DISCUSSION THREAD

**Discussion by jokulamoko** [dokku/dokku#7262] [Q&A] [answered]
(dokku version 0.35.5)

I've been struggling to deploy a Django and Celery services in one app with a Dockerfile-based Dokku deployment.

My app is built with this Dockerfile:
```
FROM python:3.10.10
...
RUN chmod +x /app/dokku/*.sh
...
ENTRYPOINT ["/app/dokku/entrypoint.sh"]
CMD ["web"]
```

```app/dokku/entrypoint.sh
#!/bin/bash
set -e

if [ "$1" = "worker" ]; then
    printf '\nStarting worker via entrypoint\n'
    exec celery -A myapp worker --loglevel=DEBUG --max-tasks-per-child=100 --optimization=fair
elif [ "$1" = "web" ]; then
    printf '\nStarting web via entrypoint\n'
    exec gunicorn --chdir /app/myapp -b :8000 myapp.wsgi:application --log-level debug --access-logfile - --error-logfile -
else
    exec "$@"
fi
```

```Procfile
web: web
worker: worker
```

My understanding is that for the push/deploy to succeed, a process needs to be running, which is why I pass CMD "web". Ideally, I don't have any CMD line in the Dockerfile, and instead use the Procfile to start both services in their respective ways, to support the completion of the deploy.

If I don't have a CMD line, when I push/deploy:
```
=====> Processing deployment checks
remote:  !     No healthchecks found in app.json for web process type
       No web healthchecks found in app.json. Simple container checks will be performed.
       For more efficient zero downtime deployments, add healthchecks to your app.json. See https://dokku.com/docs/deployment/zero-downtime-deploys/ for examples
-----> Deploying backend via the docker-local scheduler...
-----> Deploying web (count=1)
       Attempting pre-flight checks (web.1)
remote:  !     Container state: exited
673fca1989a0942eb15f44e88346a7343f0cfc1104a0a5ecca6c227a985ce419
remote:  !     Could not start due to 1 failed checks (web.1)
=====> Start of backend container output (673fca1989a0 web.1)
=====> End of backend container output (673fca1989a0 web.1)
=====> Start of backend container output (673fca1989a0 web.1)
=====> End of backend container output (673fca1989a0 web.1)
remote: parallel: This job failed:
remote: /var/lib/dokku/plugins/available/scheduler-docker-local/bin/scheduler-deploy-process-container backend dockerfile dokku/backend:latest latest web 1 1
To MY_IP_ADDRESS:backend
 ! [remote rejected] main -> master (pre-receive hook declined)
error: failed to push some refs to 'dokku@MY_IP_ADDRESS:backend'
```
Which is a fail, because the app wasn't started (neither web or worker).

However if I include the CMD line, the deployment is successful and the Procfile seemingly gets ignored - if I check the logs of the worker (celery), I get:
```
me@myapp:~$ dokku logs backend -t -p worker
2024-10-12T04:33:32.503141473Z app[worker.1]: [2024-10-12 04:33:32 +0000] [7] [INFO] Starting gunicorn 23.0.0
2024-10-12T04:33:32.505162453Z app[worker.1]: [2024-10-12 04:33:32 +0000] [7] [INFO] Listening at: http://0.0.0.0:8000 (7)
2024-10-12T04:33:32.505897819Z app[worker.1]: [2024-10-12 04:33:32 +0000] [7] [INFO] Using worker: sync
2024-10-12T04:33:32.515107271Z app[worker.1]: [2024-10-12 04:33:32 +0000] [8] [INFO] Booting worker with pid: 8
```
which is a gunicorn log, suggesting that worker was started with the "web" arg to the entrypoint.

Am I doing something wrong? Presumably this case was designed for. Why does the Dockerfile need a CMD command, when I define an entrypoint with a Procfile, per https://dokku.com/docs/processes/process-management/#procfile - specifically:
_If the app build declares an ENTRYPOINT, the command defined in the Procfile is passed as an argument to that entrypoint. This is the case for all Dockerfile-based, Docker Image, and Cloud Native Buildpack deployments._

What I've confirmed:
- The location of Procfile is correct (within the image) (DOKKU_PROCFILE)
- The docker image builds and starts successfully for both Django and Celery in local testing
- When running with CMD web, Django works fine (celery doesn't).

Ideally, I run these two as separate processes within the one app, given the image is the same for both services.

Happy to generate any versions/logs on request to help debug.

Any help would be greatly appreciated.

**Comment by jokulamoko** [comment]
Update:
- I've disabled the default healthchecks upon deploy.
- Then when I run ` dokku ps:scale backend worker=1 web=1`  then check the logs, I get nothing. So I think the containers aren't running, probably because of some issue with the Procfile not being found/understood

Any help on debugging this specific issue would be greatly appreciated :)

**Comment by jokulamoko** [comment, marked as answer]
ok I have stumbled across a solution.
1. Add DYNO vars for each service e.g.
```
dokku config:set backend DYNO_WEB=web.1
dokku config:set backend DYNO_WORKER=worker.1
```
2. Disable healthchecks (not desireable, but this is the only way you can avoid having the CMD line in your Dockerfile as far as I've worked out)
```
dokku checks:disable backend web,worker
```
3. Reference the DYNO vars in your Procfile:
```
web: DYNO=$DYNO_WEB /app/dokku/entrypoint.sh
worker: DYNO=$DYNO_WORKER /app/dokku/entrypoint.sh
```
4. Update your `entrypoint.sh` appropriately, to use the DYNO var to determine how to start the service
5. Push/deploy
6. Scale your non-web services
`dokku ps:scale backend worker=1 web=1`

This doesn't feel like an ideal pattern, I think the biggest issue is the disabling of the healthchecks. But perhaps some custom healthchecks can be defined in `app.json`, including automatic spinning up of the `worker` service - those with more expertise with Dokku can weigh in on that.

But it works! The correct commands are used to spin up each service/process in the app.

**Reply by jokulamoko** [reply]
I found this solution among the discussions in these two threads:
https://github.com/dokku/dokku/issues/1564
https://github.com/dokku/dokku/pull/1596

I would've thought a more stable pattern (e.g. Procfile overrides CMD line in Dockerfile-based builds, default healthchecks can be kept in place) would've emerged, perhaps it has and I haven't been able to find it.

**Comment by josegonzalez** [comment]
Where exactly is your Procfile located? We don't look for `DOKKU_PROCFILE` by default - the `Procfile` has to be in the root of the app code repo, and if it isn't, it is ignored.
